File: src/server/src/api/votes.go
package api

import (
	"dal"
	"database/sql"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"net/url"
	"strconv"
)

type voteRequest struct {
	Vote dal.Vote `json:"vote"`
}

type votesRequest struct {
	Votes []dal.Vote `json:"votes"`
}

func ServeVotes(w http.ResponseWriter, r *http.Request) {
	id := r.URL.Path[len("/api/votes"):]
	if len(id) > 0 {
		id = id[1:]
	}

	voteId, idErr := strconv.Atoi(id)

	db, err := dal.Open()
	if err != nil {
		log.Printf("Failed to connect to db: %s\n", err)
		http.Error(w, "failed to create vote", 500)
		return
	}
	defer db.Close()

	b, err := ioutil.ReadAll(r.Body)

	if err != nil {
		log.Printf("Failed to read body %v\n", r.Body)
		http.Error(w, "Could not read body", 400)
		return
	}

	switch r.Method {
	case "GET":
		if idErr != nil {
			findVotes(w, db, r.URL)
		} else {
			fetchVote(w, db, voteId)
		}
	case "POST":
		if idErr == nil {
			http.Error(w, "POST to /id not allowed", 405)
			return
		}
		createVote(w, db, b)
	case "PUT":
		if idErr != nil {
			http.Error(w, "PUT requires id", 405)
			return
		}
		replaceVote(w, db, b, voteId)
	case "DELETE":
		if idErr != nil {
			http.Error(w, "DELETE requires id", 405)
			return
		}
		deleteVote(w, db, voteId)
	default:
		http.Error(w, fmt.Sprintf("Method %s not recognized", r.Method), 405)
	}
}

func findVotes(w http.ResponseWriter, db *sql.DB, url *url.URL) {
	http.Error(w, "Method not implemented", 405)
}

func fetchVote(w http.ResponseWriter, db *sql.DB, id int) {
	db, err := dal.Open()

	if err != nil {
		log.Printf("Failed to open db: %s\n", err)
		http.Error(w, "Failed to connect to db", 500)
		return
	}

	defer db.Close()

	vote, err := dal.FetchVote(db, int64(id))

	if err != nil {
		http.Error(w, fmt.Sprintf("Failed to query db: %s", err), 500)
		return
	}

	resp := voteRequest{
		Vote: *vote,
	}

	j, err := json.Marshal(resp)

	if err != nil {
		http.Error(w, fmt.Sprintf("Failed to marshal to json: %v", resp), 500)
		return
	}

	w.Write(j)
}

func createVote(w http.ResponseWriter, db *sql.DB, b []byte) {
	var body voteRequest

	if err := json.Unmarshal(b, &body); err != nil {
		log.Printf("Failed to unmarshal body %s\n", b)
		http.Error(w, "Could not parse body", 400)
		return
	}

	if body.Vote.PlayerId == nil || body.Vote.AccusationId == nil || body.Vote.Accuse == nil {
		log.Printf("Player, Accusation, and Vote are required\n")
		http.Error(w, "Player, Accusation, and Vote are required", 400)
		return
	}

	vote, err := dal.CreateVote(db, &body.Vote)

	if err != nil {
		log.Printf("Failed to create vote: %s", err)
		http.Error(w, "Failed to create vote", 500)
		return
	}

	body.Vote = *vote

	j, err := json.Marshal(body)

	if err != nil {
		log.Printf("Failed to marshal to json: %v", body)
		http.Error(w, "Failed to marshal to json", 500)
		return
	}

	state, err := dal.CheckAccusationState(db, int(*vote.AccusationId))

	if state == "guilty" {
		VotedGuilty(vote.AccusationId, db)
	}

	w.Write(j)
}

func replaceVote(w http.ResponseWriter, db *sql.DB, b []byte, id int) {
	http.Error(w, "Method not implemented", 405)
}

func deleteVote(w http.ResponseWriter, db *sql.DB, id int) {
	http.Error(w, "Method not implemented", 405)
}

File: src/browser/server/mocks/votes.js
module.exports = function(app) {
  var express = require('express');
  var bodyParser = require('body-parser');
  var votesRouter = express.Router();

  var allVotes = [{
      id: '1',
      player: 2,
      accusation: 1,
      accuse: true
  }, {
      id: '2',
      player: 3,
      accusation: 1,
      accuse: false
  }, {
      id: '3',
      player: 6,
      accusation: 2,
      accuse: true
  }, {
     id: '4',
     player: 5,
     accusation: 3,
     accuse: true
  }, {
      id: '5',
      player: 7,
      accusation: 2,
      accuse: false
  }, {
      id: '6',
      player: 7,
      accusation: 3,
      accuse: true
  }, {
      id: '7',
      player: 2,
      accusation: 4,
      accuse: true
  }
  ];
  
  votesRouter.get('/', function(req, res) {
    res.send({
      'votes': allVotes.filter((function(g){
          return !req.query.state || req.query.state == g.state;
      }))
    });
  });

  votesRouter.post('/', function(req, res) {
      var body     = req.body;
      body.vote.id = 1;

      res.send(allVotes[body.vote.id]);
  });

  votesRouter.get('/:id', function(req, res) {
    res.send({
      'vote': allVotes.filter((function(g) {
          return g.id == req.params.id;
      }))[0]
    });
  });

  votesRouter.put('/:id', function(req, res) {
      var g = allVotes[0];
      g.id = req.params.id;

    res.send({
      'votes': g
    });
  });

  votesRouter.delete('/:id', function(req, res) {
    res.status(204).end();
  });

  app.use(function (req, res, next) {
    req.testing = 'testing';
    return next();
  });

  app.use('/api/votes', bodyParser.json(), votesRouter);
};
